HTT_Leavitt Normalization (Revised)
========================================================
## Author: Sumaiya Islam
## Date: June 25, 2014

### A. Set up wd & packages

Use R.2.14

Here we will continue from pre-processing of our samples to the normalization of our samples using the objects HTT_Leavitt.RData and HTT_Leavitt.2.RData.

Initially we will set our working directory and load our libraries:
```{r HTT_Leavitt_setup, include = FALSE, echo=FALSE}
setwd("/home/sislam/HTT_Leavitt")
library(methylumi)
library(gplots)
library(marray)
library(lumi)
library(lattice)
library("RColorBrewer")
library(knitr)
library(xtable)
library(wateRmelon)
library(RPMM)
library(limma)
library(sva)
```

### B. Load files

Only two files need to be loaded before we can begin:
- **HTT_Leavitt**, this is the MethylLumiM object which has been pre-processed and should contain 437,035 probes (in our particular case) and 18 samples.
- **HTT_Leavitt.2**, is the MethyLumiSet object which is needed for the normalization process and has not yet been utilized.

```{r HTT_Leavitt_fileLoad, echo=FALSE}
load("HTT_Leavitt(filtered).RData")
sampleNames(HTT_Leavitt)<-c("COB05_Cortex","HDB176_Cortex","COB_22_52_Cortex","HDB167_Cortex",    "HDB162_Cortex","COB51_Cortex","HDB119_Cortex","COB51_Liver","HDB119_Liver","HDB162_Liver",    "HDB176_Liver","HDB167_Liver","HDB166_Cortex","COB125_Cortex","COB20_30_Cortex","COB59_Cortex",     "HDB159_Cortex","HDB165_Cortex")
load("HTT_Leavitt.2.RData")
sampleNames(HTT_Leavitt.2)<-c("COB05_Cortex","HDB176_Cortex","COB_22_52_Cortex","HDB167_Cortex",    "HDB162_Cortex","COB51_Cortex","HDB119_Cortex","HDB166_Cortex","COB125_Cortex","COB20_30_Cortex", 
"COB59_Cortex","HDB159_Cortex","HDB165_Cortex","COB51_Liver","HDB119_Liver","HDB162_Liver",     "HDB176_Liver", "HDB167_Liver")
dim(HTT_Leavitt)
dim(HTT_Leavitt.2)
```
**HTT_Leavitt** should contain 437,035 probes and 18 samples where **HTT_Leavitt.2** should contain 485,577 probes and 18 samples.

### C. Density Plots

Before we get into colour correction and normalization, we'll generate a density plot of the raw data from HTT_Leavitt with the probes filtered. We will accordingly compare these raw density plots to the normalized density plots at the end of the script.

```{r HTT_Leavitt_filteredDensity, echo=FALSE}
chipCol <- rainbow(nlevels(factor(HTT_Leavitt$Sentrix_ID)))[as.numeric(factor(HTT_Leavitt$Sentrix_ID))]

density(HTT_Leavitt, col=chipCol, lty = 1, legend=NULL, main = "HTT_Leavitt after probe filtering - M-values by batch")
legend("topright",title="Sentrix ID", fill=rainbow(nlevels(factor(HTT_Leavitt$Sentrix_ID))),legend=levels(factor(HTT_Leavitt$Sentrix_ID)),cex=0.6)
```
You can see how much variability exists between the sample run across the different chips in this density plot. 

```{r HTT_Leavitt_tissueDensity, echo=FALSE}
tissueCol <- as.numeric(factor(HTT_Leavitt$Sample_Group))
tissueCol <- gsub("1", "orange",  gsub("2", "blue", tissueCol))
density(HTT_Leavitt, col=tissueCol, lty = 1, legend=NULL, main = "HTT_Leavitt after probe filtering- M-values by tissue")
legend("topleft", cex=0.6, legend = c("Cortex", "Liver"),fill = c("orange","blue"), horiz = T)
```
You can see that the different tissue types have different methylation density profiles. Moreover, the different samples within each tissue group (particularly in the liver samples) exhibits notable variability which we will try to correct through normalization. 

```{r HTT_Leavitt_SexDensity, echo=FALSE}
SexCol <- as.numeric(factor(HTT_Leavitt$Sex))
SexCol <- gsub("1", "gray", gsub("2", "deeppink",  gsub("3", "dodgerblue", SexCol)))
density(HTT_Leavitt, col=SexCol, lty = 1, legend=NULL, main = "HTT_Leavitt after probe filtering- M-values by Sex")
legend("topleft", cex = 0.5, legend = c("Female", "Male", "NA"),fill = c("deeppink","dodgerblue", "gray"), horiz = T)
```
There is some variability in the methylation density profiles due to differences in sex (but not as much as the effect of tissue differences).

### D. Subset dataset

We will subset the dataset according to the research question being asked (as provided by Rebecca). 

##### Question 1: is the HTT gene differentially methylated between HD and control cortex samples. Use all the cortex samples (remember HDB125 is now a control sample after CAG sizing; we have changed its sample name to CO125 to avoid confusion)

```{r HTT_Leavitt_subset, echo= FALSE}
# subset cortex samples
cortex.samples<- grep("Cortex", HTT_Leavitt$Sample_Group)
HTT_Leavitt.cortex<-HTT_Leavitt[,cortex.samples] # probes= 437035, n = 13
cortex.samples2<- grep("Cortex", sampleNames(HTT_Leavitt.2))
HTT_Leavitt.2.cortex<-HTT_Leavitt.2[,cortex.samples2] # probes= 485577, n = 13
sampleNames(HTT_Leavitt.cortex)
sampleNames(HTT_Leavitt.2.cortex)
```

##### Question 2: is the HTT gene differentially methylated between cortex and liver samples? Use the matched cortex and liver samples

```{r echo=FALSE}
sampleNames(HTT_Leavitt)
matched.samples<- c("HDB119_Cortex", "HDB119_Liver", "COB51_Cortex", "COB51_Liver", "HDB167_Cortex", "HDB167_Liver", "HDB176_Cortex", "HDB176_Liver", "HDB162_Cortex", "HDB162_Liver")
HTT_Leavitt.Matched<-HTT_Leavitt[,matched.samples] # probes= 437035, n = 10
HTT_Leavitt.2.Matched<-HTT_Leavitt.2[,matched.samples] # probes= 485577, n = 10
sampleNames(HTT_Leavitt.Matched)
sampleNames(HTT_Leavitt.2.Matched)
```

##### Question 3: are there global methylation changes between HD and control Cortex samples? Use all the cortex samples (remember HDB125 is now a control sample after CAG sizing; we have changed its sample name to CO125 to avoid confusion).


```{r echo=FALSE}
# same subsetted dataset as in Question 1 (see above)
sampleNames(HTT_Leavitt.cortex)
sampleNames(HTT_Leavitt.2.cortex)
```


### E. Quantile normalization

 The **lumiMethylN** is the lumi package's quantile normalization function.

```{r HTT_Leavitt_luminormalization, include=FALSE, echo=FALSE}
# all samples
HTT_Leavitt.q <- lumiMethyN(HTT_Leavitt, method = "quantile")

# cortex samples
HTT_Leavitt.cortex.q <- lumiMethyN(HTT_Leavitt.cortex, method = "quantile")

# matched samples
HTT_Leavitt.Matched.q <- lumiMethyN(HTT_Leavitt.Matched, method = "quantile")
```

### F. Normalization: SWAN vs. BMIQ

The 450K Illumina Infinium Array has inherent variation associated with its methodologies which must be accounted for in our analyses. Much of this variation is attributed to the use of two types of probes used in the array, Type I and Type II.

Type I probes contain two bead types corresponding to an unmethylated (U) or methylated (M) status. Type I probes obtain methylation status as a result of fluoresence expressed after a single base pair extension occurs just after the target basepair, resulting in only one color channel being utilized (red). Type I probes also assume that any CpG sites underlying the probe are of the same status as the query site (methylated or unmethylated). The beta values for Type I probes are then determined by this formula b= M/(U + M). Type II probes do not assume the methylation status of underlying CpG sites within the probe and so consist of a combination of degenerate probes with varying combinations of up to three underlying CpG sites. Type II probes also detect methylation status with a single base pair extension except that the site being extended is the CpG site of detection and so require two fluorescent colors green for methylated (M) and red for unmethylated (U) sites. Type II probe beta values are calculated using this formula b = Green (M)/(Red (U) + Green (M)). 

Due to these inherent differences between Type I and Type II probes used in the Illumina Infinium 450K array several groups in the field have deveolped various normalization analyses to correct for the differences between these probes. Peak-based Correction (PBC) (Dedeurwaerder *et. al* 2011 *Epigenomics*) and subset-quantile within array normalization (SWAN) (Maksimovic *et. al* 2012 *Genome Biology*) are some of the most popular methods we will be comparing to determine which provides the best normalization technique for our data (using the dmrse metric as before). We will be using only the SWAN  procedures for this dataset as the PBC method is not recommended for samples that exhibit notable intermediate methylation levels, which we observe particularly in the cortex samples.

#### Subset-quantile within array normalization (SWAN)

SWAN normalization sets out to correct for differences associated by the use of the two different probe types. As the beta value distributions vary between the two probe types, resulting from the use of different color channels, SWAN attempts to normalize the data by reducing these differences. This is accomplished by performing quantile normalization on subsets of probes stratified based on the number of underlying CpGs present within the probes. The reason for this stratification is SWAN takes into account that there are inherent biological differences between regions which are CpG dense and CpG sparse and so tries to accomplish normalizing technical variation without interferring with true biological variation (Maksimovic *et. al* 2012 *Genome Biology*). 


```{r HTT_Leavitt_SWAN, echo=FALSE}
runSWAN<- function(x,y){
  target <- fData(QCdata(x))$TargetID =="NEGATIVE"
  grn <- methylated(QCdata(x))[target,]
  red <- unmethylated(QCdata(x))[target,]
  bkg <- rowMeans(cbind(colMeans(grn),colMeans(red)))
  all(names(bkg)%in%sampleNames(y)) 
  bkg <- bkg[sampleNames(y)]
  source("SWAN scripts.R")
  load("nCpG all.RData")
  all(featureNames(y)%in%names(nCpG.all))
  fData(y)$nCpG <- nCpG.all[featureNames(y)]
  methylumi.swan <- preprocessMSWAN(y,bg=bkg)
  return(methylumi.swan)
  }

# all samples
HTT_Leavitt.swan<-runSWAN(HTT_Leavitt.2,HTT_Leavitt.q)
# checks
sum(is.na(exprs(HTT_Leavitt.swan))) # should be 0
sum(is.na(unmethylated(HTT_Leavitt.swan))) # should be 0
sum(is.na(methylated(HTT_Leavitt.swan))) # should be 0

# cortex samples
HTT_Leavitt.cortex.swan<-runSWAN(HTT_Leavitt.2.cortex,HTT_Leavitt.cortex.q)
# checks
sum(is.na(exprs(HTT_Leavitt.cortex.swan))) # should be 0
sum(is.na(unmethylated(HTT_Leavitt.cortex.swan))) # should be 0
sum(is.na(methylated(HTT_Leavitt.cortex.swan))) # should be 0

# Matched samples
HTT_Leavitt.Matched.swan<-runSWAN(HTT_Leavitt.2.Matched,HTT_Leavitt.Matched.q)
# checks
sum(is.na(exprs(HTT_Leavitt.Matched.swan))) # should be 0
sum(is.na(unmethylated(HTT_Leavitt.Matched.swan))) # should be 0
sum(is.na(methylated(HTT_Leavitt.Matched.swan))) # should be 0
```

Saving the SWAN-normalized datasets:
```{r HTT_Leavitt_SWANsave, echo=FALSE}
  # save(HTT_Leavitt.swan,file = "HTT_Leavitt.norm.RData")
  # save(HTT_Leavitt.Matched.swan,file = "HTT_Leavitt.Matched.norm.RData")
  # save(HTT_Leavitt.cortex.swan,file = "HTT_Leavitt.cortex.norm.RData")
```

### Comparing raw, quantile vs SWAN-normalized density plots

To quantitatively assess the relative performance of the different datasets produced above, we will use various performance metrics included in the wateRmelon package (Pidsley et al. 2013 *BMC Genomics*):

- **dmrse** = measures the dispersion for the beta values of 227 CpG sites located in imprinted differentially methylated regions (iDMRs). These iDMRs are expected to have approximately half-methylation levels (ie beta = 0.5) since genomic imprinting involves parent-specific mono-allelic expression of one allele with DNA methylation-based silencing of the other allele. This function returns the standard error of the mean of betas for all samples and iDMR probes.
-  **dmrse_row** calculates the between-sample standard error of the iDMR measures. 
-  **dmrse_col** calculates the between-probe standard error of the iDMR measures. 
**Note that small values indicate good performance**

Note that the other performance metrics included in the wateRmelon package such as
- **seabi**: Calculates an area under ROC curve - based metric for Illumina 450K data using a t-test for male:female difference as the predictor for X-chromosome location of probes. The metric is 1-area so
that small values indicate good performance. Note that this requires both male and female samples of known sex and can be slow to
compute due to running a t-test on every probe.
- **genki**: A very simple genotype calling by one-dimensional K-means clustering is performed on each SNP,and for those SNPs where there are three genotypes, the squared deviations are summed for each genotype (similar to a standard deviation for each of allele A homozygote, heterozygote and allele B homozygote). By default these are further divided by the square root of the number of samples to get a standard error-like statistic.
However since the XY probes and SNP probes have been filtered from our dataset, we will not be able to use seabi and genki metrics, respectively. Moreover, since we do not have technical replicates in our dataset, we cannot use the correlation of technical replicates as a performance metric of our normalization procedures, as we would usually do. 

We will show the density plots and dmrse scores for each of the datasets above:
```{r dmrse_finalobjects, echo=FALSE}
calcMetric<-function(x){
  alpha<-dmrse(x)
  beta<-dmrse_row(x)
  delta<-dmrse_col(x)
  dmrse.list<-list(dmrse = alpha, dmrse_row = beta, dmrse_col = delta)
  return(dmrse.list)
}
final.dmrse.scores<-function(a,b,c){
  c1<-calcMetric(a)
  c2<-calcMetric(b)
  c3<-calcMetric(c)
  d1<-list(c1,c2,c3)
  metric.df<-data.frame(matrix(unlist(d1), nrow=3, byrow=T))
  rownames(metric.df) <- c("Raw", "Quantile", "SWAN")
  colnames(metric.df) <- c("dmrse", "dmrse_row", "dmrse_col")
  return(metric.df)
}
```

For all samples (cortex samples are in yellow while liver samples are in blue):
```{r HTT_Leavitt_compareAll, echo=FALSE}
par(mfrow = c(1, 3))
density(HTT_Leavitt,legend=NULL, lty = 1, xlim = c(-6, 6), main = "All Raw", col=tissueCol)
density(HTT_Leavitt.q,legend=NULL, lty = 1, xlim = c(-6, 6), main = "All Quantile", col=tissueCol)
density(HTT_Leavitt.swan,legend=NULL, lty = 1, xlim = c(-6, 6), main = "All SWAN", col=tissueCol)


(all.finaldrmse<-final.dmrse.scores(betas(HTT_Leavitt), betas(HTT_Leavitt.q), betas(HTT_Leavitt.swan)))
```

For the cortex samples:
```{r HTT_Leavitt_comparecortex, echo=FALSE}
par(mfrow = c(1, 3))
density(HTT_Leavitt.cortex,legend=NULL, lty = 1, xlim = c(-6, 6), main = "Cortex Raw", col="orange")
density(HTT_Leavitt.cortex.q,legend=NULL, lty = 1, xlim = c(-6, 6), main = "Cortex Quantile", col="orange")
density(HTT_Leavitt.cortex.swan,legend=NULL, lty = 1, xlim = c(-6, 6), main = "Cortex SWAN", col="orange")

(cortex.finaldrmse<-final.dmrse.scores(betas(HTT_Leavitt.cortex), betas(HTT_Leavitt.cortex.q), betas(HTT_Leavitt.cortex.swan)))
```

For matched samples:
```{r HTT_Leavitt_compareMatched, echo=FALSE}
par(mfrow = c(1, 3))
density(HTT_Leavitt.Matched,legend=NULL, lty = 1, xlim = c(-6, 6), main = "Matched Raw", col="blue")
density(HTT_Leavitt.Matched.q,legend=NULL, lty = 1, xlim = c(-6, 6), main = "Matched Quantile", col="blue")
density(HTT_Leavitt.Matched.swan,legend=NULL, lty = 1, xlim = c(-6, 6), main = "Matched SWAN", col="blue")

(Matched.finaldrmse<-final.dmrse.scores(betas(HTT_Leavitt.Matched), betas(HTT_Leavitt.Matched.q), betas(HTT_Leavitt.Matched.swan)))
```

We can see that there is a dramatic reduction in sample-to-sample variability from the raw data to SWAN-normalized in all the datasets. Oddly, this is not reflected by the the dmrse scores as the raw datasets exhibit slightly better (ie lower scores) between-sample standard error for the iDMR measures (dmrse_row) for the datasets.

### G. Heatmaps

For all samples:
```{r HTT_Leavitt_allSWANheatmap, fig.width=9, fig.height=8, echo=FALSE}
tissueCol <- as.numeric(factor(HTT_Leavitt.swan$Sample_Group))
tissueCol <- gsub("1", "orange",  gsub("2", "blue", tissueCol))
SexCol <- as.numeric(factor(HTT_Leavitt.swan$Sex))
SexCol <- gsub("1", "gray", gsub("2", "deeppink",  gsub("3", "dodgerblue", SexCol)))

BuPu <- colorRampPalette(brewer.pal(n = 9, "BuPu"))

SWANcor <- cor(betas(HTT_Leavitt.swan), use = "pairwise.complete.obs")

heatmap.2(SWANcor,main = "Correlation of all samples - Normalized dataset",trace = "none", col = BuPu, dendrogram = "both",
          RowSideColors = SexCol, cexRow = 1,
          ColSideColors = tissueCol, cexCol = 1,
          keysize = 1)
legend("topright", bty="n", legend = c("Cortex", "Liver"),fill = c("orange","blue"), horiz = T)
legend("top", legend = c("Female", "Male", "NA"),fill = c("deeppink","dodgerblue", "gray"), horiz = T)
```

For all samples:
```{r final_heatmap_SenTiss.all, fig.width=9, fig.height=8, echo=FALSE}
tissueCol <- as.numeric(factor(HTT_Leavitt.swan$Sample_Group))
tissueCol <- gsub("1", "orange",  gsub("2", "blue", tissueCol))
sentrixCol <- HTT_Leavitt.swan$Sentrix_ID
sentrixCol <- gsub("8221916079", "lightslateblue", gsub("8221932102",  "darkgreen", sentrixCol))

BuPu <- colorRampPalette(brewer.pal(n = 9, "BuPu"))

cor <- cor(betas(HTT_Leavitt.swan), use = "pairwise.complete.obs")

heatmap.2(cor,main = "Correlation of samples - Normalized dataset",
          trace = "none", col = BuPu, dendrogram = "both",
          RowSideColors = sentrixCol, cexRow = 1,
          ColSideColors = tissueCol, cexCol = 1,
          keysize = 1)
legend("topright", bty="n", legend = c("Cortex", "Liver"),fill = c("orange","blue"), horiz = T)
legend("top", bty = "n", title = "Sentrix ID", c("8221916079", "8221932102"), cex = 1, fill = c("lightslateblue","darkgreen"), horiz=T)
```

For cortex samples:
```{r HTT_Leavitt_cortexSWANheatmap, fig.width=9, fig.height=8, echo=FALSE}

SexCol.cortex <- as.numeric(factor(HTT_Leavitt.cortex.swan$Sex))
SexCol.cortex <- gsub("1", "gray", gsub("2", "deeppink",  gsub("3", "dodgerblue", SexCol.cortex)))

BuPu <- colorRampPalette(brewer.pal(n = 9, "BuPu"))

SWAN.cortex.cor <- cor(betas(HTT_Leavitt.cortex.swan), use = "pairwise.complete.obs")

heatmap.2(SWAN.cortex.cor,main = "Correlation of cortex samples - Normalized dataset",trace = "none", col = BuPu, dendrogram = "both",
          cexRow = 1,
          ColSideColors = SexCol.cortex, cexCol = 1,
          keysize = 1)
legend("top", legend = c("Female", "Male", "NA"),fill = c("deeppink","dodgerblue", "gray"), horiz = T)
```

For matched samples:
```{r HTT_Leavitt_liverSWANheatmap, fig.width=9, fig.height=8, echo=FALSE}
tissueCol <- as.numeric(factor(HTT_Leavitt.Matched.swan$Sample_Group))
tissueCol <- gsub("1", "orange",  gsub("2", "blue", tissueCol))
SexCol <- as.numeric(factor(HTT_Leavitt.Matched.swan$Sex))
SexCol <- gsub("1", "gray", gsub("2", "deeppink",  gsub("3", "dodgerblue", SexCol)))

BuPu <- colorRampPalette(brewer.pal(n = 9, "BuPu"))

SWANcor <- cor(betas(HTT_Leavitt.Matched.swan), use = "pairwise.complete.obs")

heatmap.2(SWANcor,main = "Correlation of matched samples - Normalized dataset",trace = "none", col = BuPu, dendrogram = "both",
          RowSideColors = SexCol, cexRow = 1,
          ColSideColors = tissueCol, cexCol = 1,
          keysize = 1)
legend("topright", bty="n", legend = c("Cortex", "Liver"),fill = c("orange","blue"), horiz = T)
legend("top", legend = c("Female", "Male", "NA"),fill = c("deeppink","dodgerblue", "gray"), horiz = T)
```

### H. Plot beta values before/after normalization

We will show the distribution of beta values for the entire dataset (all samples normalized together) in its raw form and final normalized form.

```{r HTT_Leavitt_betaPlot, echo=FALSE} 
type <- fData(HTT_Leavitt.q)$INFINIUM_DESIGN_TYPE

before.betas <- betas(HTT_Leavitt)
after.betas.SWAN <- betas(HTT_Leavitt.swan)

par(mfrow = c(1, 1))
plot(c(0, 1) , c(0, 10), type = "n", xlab = "Beta values", ylab = "Density")
lines(density(na.omit(before.betas)[type=="I",]),col="blue",lty=1)
lines(density(na.omit(before.betas)[type=="II",]),col="blue",lty=2, wld=1.5)
lines(density(na.omit(after.betas.SWAN)[type=="I",]),col="red",lty=1)
lines(density(na.omit(after.betas.SWAN)[type=="II",]),col="red",lty=2, lwd=1.5)
legend("top", inset=.05, cex=0.6, c("Raw Type 1","Raw Type 2", "Normalized Type 1", "Normalized Type 2"), col=c("blue","blue","red", "red"), lty=c(1,2,1,2), horiz=FALSE)
```

We can see that there is improved peak-to-peak overlap of the Type 1 and Type 2 probes with the final normalized dataset as desired (ie Type 1 and 2 probes have more similar dynamic ranges for their beta values). 

We will continue on with our analysis using the SWAN-normalized datasets.